lib/lita/handlers/stacker.rb

```diff
@@ -1,53 +1,119 @@
+# frozen_string_literal: true
+
 module Lita
   module Handlers
     class Stacker < Handler
-      route %r{^stack(\s+\@\p{Word}+\s*)?$}, :lifo_add
-      route %r{^unstack(\s+\@\p{Word}+\s*)?$}, :lifo_remove
-      route %r{^stack drop}, :lifo_remove
-      route %r{^stack show}, :lifo_peek
-      route %r{^stack clear}, :lifo_clear
+      VERSION = '0.2.0'
+
+      route(/^stack(\s+(on.*|\@\p{Word}+\s*))?$/, :lifo_add, help: {
+              t('add.help.simple.command') => t('add.help.simple.description'),
+              t('add.help.on.command') => t('add.help.on.description'),
+              t('add.help.another.command') => t('add.help.another.description')
+            })
+      route(/^(unstack|stack (drop|done))(\s+\@\p{Word}+\s*)?$/, :lifo_remove, help: {
+              t('remove.help.unstack.command') => t('remove.help.unstack.description'),
+              t('remove.help.done.command') => t('remove.help.done.description'),
+              t('remove.help.drop.command') => t('remove.help.drop.description')
+            })
+      route(/^stacks? show/, :lifo_peek, help: {
+              t('peek.help.singular.command') => t('peek.help.singular.description'),
+              t('peek.help.plural.command') => t('peek.help.plural.description')
+            })
+      route(/^stacks? clear/, :lifo_clear, help: {
+              t('clear.help.singular.command') => t('clear.help.singular.description'),
+              t('clear.help.plural.command') => t('clear.help.plural.description')
+            })
 
       def lifo_add(response)
-        return if is_incompatible?(response)
+        return if incompatible?(response)
+
+        # There is no LSCAN Redis command, so we can hack one.
+        # I am pretty sure Lita is synchronous so it is not possible for this
+        # to be inconsistent. But in the iterest of robustness, run it as a
+        # Lua script, to keep it atomic.
+        script = <<~LUA
+          for i=0,(redis.call('llen', KEYS[1])-1) do
+            if (redis.call('lindex', KEYS[1], i) == ARGV[1]) then
+              return -1
+            end
+          end
+          redis.call('rpush', KEYS[1], ARGV[1])
+          return redis.call('lindex', KEYS[1], -2)
+        LUA
```

Comment:
I don't understand the need for this lua script, and the script below (which is there for the same reasons, I presume?).  It's possible I'm just insufficiently schooled in the ways of redis and atomicity therein, but it's also possible that others who tread here might be in a similar state.

So I wonder: Is the extra robustness here worth possibly confusing future code spelunkers?

Reply:
Some discussion of atomicity can be found [here](https://redis.io/commands/eval#atomicity-of-scripts).

You concerns are well founded, since most folks coming to this will not know Lua (I didn’t). However, I want to counter with 2 thoughts:

1. Given the perniciousness and difficulty in debugging concurrency bugs, I generally err on the side of making things as concurrently resilient as possible. Especially because concurrent access can be added at any time without realizing there are side-effects to consider.
1. Speaking of far-flung future improvements. I’m still thinking the next batch will include changing the underlying data type to be a sorted-set, so that stacks can expire after _<n>_ hours, and we don’t have long lists of stacks left over from previous meetings. A lot of the hoop-jumping in Lua is necessitated by missing Redis commands for Lists, so we can revisit the need for Lua when the data type changes.